scripts/tests/test_api_report.py

```diff
@@ -0,0 +1,326 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+"""
+Tests the api_report module.
+
+Test metadata Yaml files are stored in the test_api_report_yamls subfolder.
+"""
+
+from datetime import date
+import os
+from unittest.mock import patch, mock_open
+
+# pylint: disable=redefined-outer-name
+import pytest
+
+import api_report
+
+
+def mock_walk(examples_folder):
+    """Mock the os.walk function to return a single file."""
+    folder_list = []
+    file_list = ["metadata.yaml"]
+    return [("", folder_list, file_list)]
+
+
+def verify_example_data(example):
+    """Verifies basic correctness of example data."""
+    if 'description' in example:
+        assert isinstance(example['description'], str)
+    assert isinstance(example['created'], date)
+    assert len(example['files']) >= 1
+    file = example['files'][0]
+    assert isinstance(file['path'], str)
+    name, ext = os.path.splitext(file['path'])
+    assert name
+    assert ext.lstrip('.') in api_report.EXT_LOOKUP
+    assert len(file['apis']) >= 1
+    api = file['apis'][0]
+    assert isinstance(api['service'], str)
+    assert len(api['operations']) >= 1
+    assert isinstance(api['operations'][0], str)
+
+
+def test_gather_data_complex(monkeypatch):
+    """Reads complex metadata and verifies the output."""
+    def mock_join(folder_name, file_name):
+        return "test_api_report_yamls/complex_metadata.yaml"
+
+    monkeypatch.setattr(os, "walk", mock_walk)
+    monkeypatch.setattr(os.path, "join", mock_join)
```

Comment:
No blocker here but another way to write this could be:
```
mock_join = ''test_api_report_yamls/complex_metadata.yaml"
monkeypatch.setattr(os.path, ''join", Mock(return_value=mock_join))
```

Reply:
Makes sense, but I left this one as is for consistency with the way I'm using mock_walk.